### PR TITLE
add google analytics to flutter_tools

### LINKED ISSUE #1065: flutter command should have analytics
We should have analytics reporting of common usage from the flutter command to we can make sure it's correctly serving users.

@sethladd has experience doing this in the past.

### PR DESCRIPTION
**Pull request by devoncarew**
This adds support for sending anonymous usage data for flutter_tools to google analytics. It:
- prints a message on first run, telling the user how to disable the analytics, and displaying the privacy policy
- adds a new `flutter config` command, where analytics can be enabled and disabled
- sends command usage, command timing, and crash data

We'll want to iterate on it to properly track the host OS and command parameters.

fix #1065

### CODE REVIEW

**Comment by sethladd** [comment]
>  and command parameters

Do you mean "multi-arg commands" like `flutter build apk` ?

File: packages/flutter_tools/lib/executable.dart [hunk 1]
```diff
@@ -82,25 +84,35 @@ Future<Null> main(List<String> args) async {
     context[DeviceManager] = new DeviceManager();
     Doctor.initGlobal();
 
+    if (flutterUsage.isFirstRun) {
+      printStatus(
+        'The Flutter tool anonymously reports feature usage statistics and basic crash reports to improve\n'
+        'the tool over time; use "flutter config" to control this behavior. See our privacy policy:\n'
+        'https://www.google.com/intl/en/policies/privacy/.\n'
```
Comment (Hixie) [inline packages/flutter_tools/lib/executable.dart]: remove trailing `.`

File: packages/flutter_tools/lib/executable.dart [hunk 2]
```diff
@@ -82,25 +84,35 @@ Future<Null> main(List<String> args) async {
     context[DeviceManager] = new DeviceManager();
     Doctor.initGlobal();
 
+    if (flutterUsage.isFirstRun) {
+      printStatus(
+        'The Flutter tool anonymously reports feature usage statistics and basic crash reports to improve\n'
+        'the tool over time; use "flutter config" to control this behavior. See our privacy policy:\n'
```
Comment (Hixie) [inline packages/flutter_tools/lib/executable.dart]: "our" is unclear here. This is Google's privacy policy, not "the Flutter project's". Maybe rephrase this as:

```
The Flutter tool anonymously reports feature usage statistics and basic crash reports to Google to
help Google contribute improvements to Flutter over time. Use "flutter config" to control this behavior.
See Google's privacy policy: https://www.google.com/intl/en/policies/privacy/
```

File: packages/flutter_tools/lib/src/commands/config.dart [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import '../globals.dart';
+import '../runner/flutter_command.dart';
+
+class ConfigCommand extends FlutterCommand {
+  ConfigCommand() {
+    argParser.addFlag('analytics',
+      negatable: true,
+      help:
+        'Enable or disable reporting anonymously tool usage statistics and basic crash reports\n'
+        'to Google Analytics. See our privacy policy: www.google.com/intl/en/policies/privacy.');
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/config.dart]: s/our/Google's/

**Comment by Hixie** [comment]
I think it is critical that we add a test that verifies that if you turn off analytics, we don't send any analytics.

File: packages/flutter_tools/lib/src/usage.dart [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    return new UsageTimer._(event, _ga.startTimer(event));
+  }
+
+  void sendException(dynamic exception, StackTrace trace) {
+    String message = '${exception.runtimeType}; ${sanitizeStacktrace(trace)}';
+    printTrace('usage: $message');
+    _ga.sendException(message);
+  }
+
+  Future<Null> waitForLastPing() {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/usage.dart]: this name is confusing. I know it's just the same as the underlying API's, but I think we should name it something clearer.

**Comment by Hixie** [comment]
LGTM if, and only if, we have a test that verifies that once you opt-out, we don't send analytics. Ideally this would test the following scenarios:
* flutter create in two different directories, opt-out while in one, are you opted out of the other?
* opt out then flutter upgrade (not sure how you simulate this usefully), are you still opted out?
* opt out then change channel (alpha -> master, say), are you still opted out?

**Comment by Hixie** [comment]
Also, please update our docs so that when they introduce the flutter tool, they mention how to opt-out.

File: packages/flutter_tools/lib/src/usage.dart [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  /// Returns [Usage] active in the current app context.
+  static Usage get instance => context[Usage] ?? (context[Usage] = new Usage());
+
+  Usage() {
+    _ga = new AnalyticsIO(_kFlutterUA, 'flutter', FlutterVersion.getVersionString());
+
+    // Check if this is the first run. If so, enable analytics. We show opt-out
+    // methods (flutter config) the first time the tool is run.
+    if (!_ga.hasSetOptIn) {
+      _isFirstRun = true;
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/usage.dart]: can we avoid sending analytics on the run where we give the opt-out information?

**Comment by devoncarew** [comment]
> Do you mean "multi-arg commands" like flutter build apk?

No, I meant things like `--foo` - flags and options to commands. `flutter build apk` would come in as `apk`. For the nested commands, we could manually override to something like `build_apk`.

File: packages/flutter_tools/lib/executable.dart (hunk 2 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/executable.dart]: updated

File: packages/flutter_tools/lib/executable.dart (hunk 1 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/executable.dart]: It's there to complete the sentence :)

File: packages/flutter_tools/lib/src/commands/config.dart (hunk 3 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/config.dart]: updated

File: packages/flutter_tools/lib/src/usage.dart (hunk 5 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/usage.dart]: Yup; I'll update to not track the `config` command.

File: packages/flutter_tools/lib/src/usage.dart (hunk 4 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/usage.dart]: `ensureAnalyticsSent()`?

**Comment by devoncarew** [comment]
I'll add some tests and update the PR. We store the opt-out data in `~/.flutter`, so it would work independent of the current project, or something like running `flutter upgrade`.

File: packages/flutter_tools/lib/src/usage.dart (hunk 5 above)
Comment (Hixie) [inline packages/flutter_tools/lib/src/usage.dart]: I meant on the first run. The one where we say how to opt-out. But yeah, not tracking the config command seems like an excellent idea also!

**Comment by devoncarew** [comment]
(I'm going to rev. the `usage` package to make this easier to test.)

**Comment by devoncarew** [comment]
@Hixie, this PR has been updated quite a bit since the first review. It doesn't send any analytics on first; it doesn't send any analytics for the `config` command (in case the user is in the process of disabling analytics); we print the analytics state in the help for `config` now; we rev'd to `2.2.0` of usage to capture some changes that make testing easier; and there are two tests to verify that we don't send data when we shouldn't.

File: packages/flutter_tools/lib/executable.dart [hunk 6]
```diff
@@ -82,25 +84,35 @@ Future<Null> main(List<String> args) async {
     context[DeviceManager] = new DeviceManager();
     Doctor.initGlobal();
 
+    if (flutterUsage.isFirstRun) {
+      printStatus(
+        'The Flutter tool anonymously reports feature usage statistics and basic crash reports to Google to\n'
+        'help Google contribute improvements to Flutter over time. Use "flutter config" to control this\n'
+        'behavior. See Google\'s privacy policy: https://www.google.com/intl/en/policies/privacy/.\n'
```
Comment (Hixie) [inline packages/flutter_tools/lib/executable.dart]: remove trailing "." in url

File: packages/flutter_tools/lib/executable.dart [hunk 7]
```diff
@@ -159,3 +171,17 @@ String _doctorText() {
...
+Future<Null> _exit(int code) async {
+  if (flutterUsage.enabled) {
+    Stopwatch stopwatch = new Stopwatch()..start();
+    await flutterUsage.ensureAnalyticsSent();
+    printTrace('ensureAnalyticsSent: ${stopwatch.elapsedMilliseconds}ms');
+  }
+
+  logger.flush();
+
+  await Timer.run(() {
```
Comment (Hixie) [inline packages/flutter_tools/lib/executable.dart]: add a comment here explaining the hoops we're jumping through, in case someone looks at this code and cargo-cults it...

File: packages/flutter_tools/lib/src/commands/config.dart [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+import '../globals.dart';
+import '../runner/flutter_command.dart';
+
+class ConfigCommand extends FlutterCommand {
+  ConfigCommand() {
+    String usageStatus = flutterUsage.enabled ? 'enabled' : 'disabled';
+
+    argParser.addFlag('analytics',
+      negatable: true,
+      help: 'Enable or disable reporting anonymously tool usage statistics and crash reports ($usageStatus).');
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/config.dart]: `(currently $usageStatus)` to make it clear whether the parenthetical is the current state or the state that you'll set if you call it

File: packages/flutter_tools/lib/src/commands/config.dart [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  @override
+  final String name = 'config';
+
+  @override
+  final String description =
+    'Configure Flutter settings.\n\n'
+    'The Flutter tool anonymously reports feature usage statistics and basic crash reports to help improve\n'
+    'Flutter tools over time. See Google\'s privacy policy: www.google.com/intl/en/policies/privacy.';
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/config.dart]: remove trailing period

File: packages/flutter_tools/lib/src/commands/config.dart [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+  @override
+  final List<String> aliases = <String>['configure'];
+
+  @override
+  bool get requiresProjectRoot => false;
+
+  /// Return `null` to disable tracking of the `config` command.
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/config.dart]: presumably some commands have it disabled (because they don't override it)... I think it might be clearer if it was abstract here and we explicitly had to make a choice for each command.

File: packages/flutter_tools/lib/src/usage.dart [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+  /// Fires whenever analytics data is sent over the network; public for testing.
+  Stream<Map<String, dynamic>> get onSend => _analytics.onSend;
+
+  /// Returns when the last analytics event has been sent, or after a fixed
+  /// (short) delay, whichever is less.
+  Future<Null> ensureAnalyticsSent() {
+    // TODO(devoncarew): This may delay tool exit and could cause some analytics
+    // events to not be reported. Perhaps we could send the analytics pings
+    // out-of-process from flutter_tools?
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/usage.dart]: I think what we should do here is track pings that aren't successfully sent, and then send them the next time the tool is started, asynchronously. However, we'd have to give each ping a globally unique ID so that we could de-dupe on the server in the cases where we think it wasn't sent but it actually was.

File: packages/flutter_tools/test/analytics_test.dart [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+      DoctorCommand doctorCommand = new DoctorCommand();
+      runner = createTestCommandRunner(doctorCommand);
+      code = await runner.run(<String>['doctor']);
+      expect(code, equals(0));
+      expect(count, 0);
+    }, overrides: <Type, dynamic>{
+      Usage: new Usage()
+    });
+
+    // Ensure we con't send for the 'flutter config' command.
```
Comment (Hixie) [inline packages/flutter_tools/test/analytics_test.dart]: "don't"

File: packages/flutter_tools/lib/src/usage.dart (hunk 11 above)
Comment (Hixie) [inline packages/flutter_tools/lib/src/usage.dart]: (not in this PR, though)

**Comment by Hixie** [comment]
LGTM

File: packages/flutter_tools/lib/executable.dart (hunk 6 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/executable.dart]: done

File: packages/flutter_tools/lib/executable.dart (hunk 7 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/executable.dart]: kk

File: packages/flutter_tools/lib/src/commands/config.dart (hunk 8 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/config.dart]: done

File: packages/flutter_tools/lib/src/commands/config.dart (hunk 9 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/config.dart]: done

File: packages/flutter_tools/lib/src/commands/config.dart (hunk 10 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/config.dart]: Just one - `config`. The default (defined in `FlutterCommand`) is to return the name of the command. `config` is the only one we don't want to track.

File: packages/flutter_tools/test/analytics_test.dart (hunk 12 above)
Comment (devoncarew) [inline packages/flutter_tools/test/analytics_test.dart]: done

File: packages/flutter_tools/lib/src/usage.dart (hunk 11 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/usage.dart]: We don't have control of the backend - it's just standard google analytics. But I have seen tools that do something similar to what you're describing. Any un-set commands are buffered to a file for sending during the next session.

Analytics is generally a best effort thing however; if we lose some (small) portion of our pings, that's ok. It's more a problem when we lose some type preferentially, and end up skewing the data.

File: packages/flutter_tools/lib/src/commands/config.dart (hunk 10 above)
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/config.dart]: oh, i see, i misunderstood where i was when reading the code. don't mind me.

File: packages/flutter_tools/lib/src/usage.dart (hunk 11 above)
Comment (Hixie) [inline packages/flutter_tools/lib/src/usage.dart]: roger
